File: Demo/DemoVisualBasic/Shapes/Rectangle.vb

Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace Demo.Shapes
	Class Rectangle
		Inherits Shape

		Private _Point1 As Point
		Private _Point2 As Point

		Protected Sub New()
		End Sub

		Public Sub New(p1 As Point, p2 As Point)
			Me.Point1 = p1
			Me.Point2 = p2
		End Sub

		Public Property Point1() As Point
			Get
				Return _Point1
			End Get
			Protected Set
				_Point1 = Value
			End Set
		End Property

		Public Property Point2() As Point
			Get
				Return _Point2
			End Get
			Protected Set
				_Point2 = Value
			End Set
		End Property

		Public Overrides ReadOnly Property Center() As Point
			Get
				Dim xs = Me.Point1.X + Me.Point2.X
				Dim ys = Me.Point1.Y + Me.Point2.Y

				Return New Point(xs / 2, ys / 2)
			End Get
		End Property

		Public ReadOnly Property Width() As Integer
			Get
				Return Math.Abs(Me.Point1.X - Me.Point2.X)
			End Get
		End Property

		Public ReadOnly Property Height() As Integer
			Get
				Return Math.Abs(Me.Point1.Y - Me.Point2.Y)
			End Get
		End Property

		Public Overrides Sub Print(output As TextWriter)
			output.WriteLine($"{Me.Color} rectangle at {Me.Center} of width {Me.Width} and height {Me.Height}")
		End Sub
	End Class
End Namespace

File: Demo/DemoCSharp/Shapes/Rectangle.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Demo.Shapes
{
	class Rectangle : Shape
	{
		protected Rectangle()
		{
		}

		public Rectangle(Point p1, Point p2)
		{
			this.Point1 = p1;
			this.Point2 = p2;
		}

		public Point Point1 { get; protected set; }
		public Point Point2 { get; protected set; }

		public override Point Center
		{
			get
			{
				var xs = this.Point1.X + this.Point2.X;
				var ys = this.Point1.Y + this.Point2.Y;

				return new Point(xs / 2, ys / 2);
			}
		}

		public int Width
		{
			get { return Math.Abs(this.Point1.X - this.Point2.X);  }
		}

		public int Height
		{
			get { return Math.Abs(this.Point1.Y - this.Point2.Y); }
		}

		public override void Print(TextWriter output)
		{
			output.WriteLine($"{this.Color} rectangle at {this.Center} of width {this.Width} and height {this.Height}");
		}
	}
}
